tests/unit/utils.test.js

```diff
@@ -48,6 +63,139 @@ describe('Testing isHexWithGivenByteSize', () => {
     expect(isHexWithGivenByteSize('0x1234', 2)).toBe(true);
     expect(isHexWithGivenByteSize('0x1234ef', 3)).toBe(true);
   });
+
+  const crateElapsedTimeSec = () => {
+    const start = +new Date();
+
+    return (expected) => expect(Math.round((+new Date() - start) / 1e2) / 10).toBe(expected);
+  };
+
+  test('`timeout` works properly', async () => {
+    const expectElapsedTimeSec = crateElapsedTimeSec();
+
+    await Promise.all(
+      [
+        async () => {
+          expect(await timeout(1e2)).toBe(void 0);
+          expectElapsedTimeSec(0.1);
+        },
+        async () => {
+          expect(await timeout(3e2, () => 'a')).toBe('a');
+          expectElapsedTimeSec(0.3);
+        },
+        async () => {
+          await expect(() => timeout(2e2, () => {
+            throw new Error('Rejected timeout');
+          })).rejects.toThrowErrorMatchingSnapshot();
+          expectElapsedTimeSec(0.2);
+        },
+      ].map((f) => f()),
+    );
+  });
+
+  test('`ReusablePromiseMap` works properly', async () => {
+    const map = new ReusablePromiseMap();
+
+    const results = await Promise.all([
+      map.callByKey(
+        1,
+        () => timeout(5e2, () => 10),
+      ),
+      timeout(2e2, () => map.callByKey(1, () => Promise.resolve(2))),
+      timeout(7e2, () => map.callByKey(1, () => Promise.resolve(1))),
+    ]);
+
+    expect(results).toEqual([10, 10, 1]);
+
+    await expect(() => map.callByKey(10, () => Promise.reject(1))).rejects;
+    expect(map.map.size).toBe(0);
+  });
+
+  test('`ReusablePromiseMap` capacity', async () => {
+    const map = new ReusablePromiseMap({ capacity: 2 });
+
+    const expectElapsedTimeSec = crateElapsedTimeSec();
+
+    const results = await Promise.all(Array.from({ length: 10 }, (_, i) => map.callByKey(i, () => timeout(5e2, () => i))));
```

Comment:
`ReusablePromiseMap` has a capacity of 2 items, and because we have 10 promises being executed over different keys, `queue` will be used after the first two items are pushed. Also, check out that time will be verified by the end of the call to ensure that we execute no more than 2 promises at the same time.